### PR TITLE
spec(core): spec/impl for transform and vars 🙀

### PR DESCRIPTION
**Pull request by srl295**
Includes:
- #8966
- #8942 
- #8943 

core:
- spec updates for transforms `tran`/`bksp` and variables `vars`, removed `finl` and folded `ordr` into subtable of `tran`/`bksp`

common
- pattern-parser objects for parsing variables and regex
- KMX+ updates to support above
- utility functions on `Vars` to support variable operations
- utility functions on messages for formatting subexceptions

developer:
- refactor compiler system make compilation async
- refactor compiler sections to add async init (for unicodeset bringup)
- refactor compiler sections to have a dependency system
- make `strs`/`elem`/`list` actual compilers ('empty compilers' since they take no input)
- fixes to test harnesses to support the above
- `vars` and `tran`/`bksp` writing/rewriting
- builder for `vars`, and initial builder for `tran`/`bksp` (not fully proven)

For: #7377

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by srl295** [comment]
- might need some additional fields to preserve the pre-processed XML values.
- I also thought about having a table with all of the markers , it could probably be a single list.  Might do it that way, to keep track of them:

   - pre-scan all XML and find `\m{grave}` and `\m{abc}`

   - just put them into a simple `list` ( `abc`, `grave`… ) but where the position is significant:

| # | marker | pua |
|---|--------|-----|
| 0 | abc | E000 |
| 1 | grave | E001 |
…

**Comment by srl295** [comment]
> We probably cannot safely use PUA for markers, because keyboards will exist that use PUA. In Keyman, we've used U+FFFF as a sentinel for special codes, such as markers. For example, markers are stored as: (UC_SENTINEL, CODE_DEADKEY, 1-based-marker-id).

We'll have to scan the output text anyway to remove markers or turn them into context actions.

Could turn around that scheme a little, and do something like this:

U+E000 PUA-escape  ( U+FFFF is a noncharacter) 
U+E001 PUA-marker #1
U+E002 PUA-marker #2 etc.

Then U+E001 => marker 1, but U+E000 U+E001  = U+E001, and U+E000 U+E000 = U+E000 


> 
> Apart from that, a pre-scan to find all markers sounds like an excellent plan. (Or could allocate on first use?)

**Comment by srl295** [comment]
swapped the PRs around (this one's a draft!) so that this one could depend on uset

**Comment by mcdurdin** [comment]
> U+E000 PUA-escape ( U+FFFF is a noncharacter)

We currently use U+FFFF internally precisely because it _is_ a noncharacter, and can safely assume that it never needs to be in text (in fact, we can strip it out if it does appear). Keeps plain text plain and allows us to process just the markers specially.

**Comment by srl295** [comment]
ok 🙀  i've skipped the e2e test- it shows prominently as skipped- this still won't pass all tests, but I'd like to get reviews on it before it gets even more divergent.

File: common/web/types/src/kmx/kmx-plus-builder/build-tran.ts [hunk 1]
```diff
@@ -1,50 +1,90 @@
...
+        mapFrom: build_strs_index(sect_strs, transform.mapFrom),
+        mapTo: build_strs_index(sect_strs, transform.mapTo),
+      });
+    }
+    for (let reorder of group.reorders) {
+      tran.reorders.push({
+        elements: build_elem_index(sect_elem, reorder.elements),
+        before: build_elem_index(sect_elem, reorder.before),
+      });
+    }
```
Comment (srl295) [inline common/web/types/src/kmx/kmx-plus-builder/build-tran.ts:78]: bugs here, see https://github.com/keymanapp/keyman/pull/8943

**Comment by mcdurdin** [comment]
Build failing with (e.g. on Windows):

```
07:47:14   ..\..\..\src\kmx/kmx_plus.h(225): error C2065: 'LDML_SECTIONID_ORDR': undeclared identifier
07:47:14   ..\..\..\src\kmx/kmx_plus.h(225): error C2131: expression did not evaluate to a constant
07:47:14   ..\..\..\src\kmx/kmx_plus.h(225): note: a non-constant (sub-)expression was encountered
07:47:14   ..\..\..\src\kmx/kmx_plus.h(236): error C2065: 'LDML_LENGTH_ORDR': undeclared identifier
07:47:14   ..\..\..\src\kmx/kmx_plus.h(300): error C2338: mismatched size of section tran
```

**Comment by srl295** [comment]
> Build failing with (e.g. on Windows):
> 
> 
> 
> ```
> 
> 07:47:14   ..\..\..\src\kmx/kmx_plus.h(225): error C2065: 'LDML_SECTIONID_ORDR': undeclared identifier
> 
> 07:47:14   ..\..\..\src\kmx/kmx_plus.h(225): error C2131: expression did not evaluate to a constant
> 
> 07:47:14   ..\..\..\src\kmx/kmx_plus.h(225): note: a non-constant (sub-)expression was encountered
> 
> 07:47:14   ..\..\..\src\kmx/kmx_plus.h(236): error C2065: 'LDML_LENGTH_ORDR': undeclared identifier
> 
> 07:47:14   ..\..\..\src\kmx/kmx_plus.h(300): error C2338: mismatched size of section tran
> 
> ```

Fixed in later pr - the seaside

**Comment by srl295** [comment]
@mcdurdin  would like to merge this and the child PRs once #8943  passes tests which I expect shortly!
I'm assuming it's OK to merge this and children since they'll then all build…

File: common/web/types/src/kmx/element-string.ts [hunk 2]
```diff
@@ -24,14 +24,19 @@ export class ElemElement {
...
       return;
     }
 
-    let items = source.split("");
+    let items;
+    if (Array.isArray(source)) {
+      items = source;
+    } else {
+      items = source.split("");
+    }
```
Comment (mcdurdin) [inline common/web/types/src/kmx/element-string.ts:39]:
```suggestion
    let items = Array.isArray(source) ? source : source.split("");
```

Less verbose but just a nit grin

File: common/web/types/src/kmx/kmx-plus-builder/build-vars.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+* Builder for the 'vars' section
+*/
+export function build_vars(kmxplus: KMXPlusData, sect_strs: BUILDER_STRS /*, sect_elem: BUILDER_ELEM*/) : BUILDER_VARS {
+  if(!kmxplus.vars) {
+    return null;
+  }
+
+  const stringVars = kmxplus.vars.strings.map(v => <BUILDER_VARS_ITEM>{
```
Comment (mcdurdin) [inline common/web/types/src/kmx/kmx-plus-builder/build-vars.ts:28]: sneaky but nice way of avoiding `v => { return { ... } }`?

File: common/web/types/src/ldml-keyboard/pattern-parser.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+/**
+ * Class for helping with markers
+ */
+export class VariableParser {
+  /**
+   * A marker id has the same constraint as a key id. TODO-LDML: Needs to be reflected in the spec
+   */
+  public static readonly ID = /^[0-9A-Za-z_]{1,32}$/;
```
Comment (mcdurdin) [inline common/web/types/src/ldml-keyboard/pattern-parser.ts:58]: This regex is the same as for MarkerParser. Can we consolidate?

File: common/web/types/src/util/compiler-interfaces.ts [hunk 5]
```diff
@@ -44,6 +44,21 @@ export function compilerErrorFormatCode(code: number): string {
...
+/**
+ * @param e event or array of events
+ * @returns string
+ */
+export function compilerEventFormat(e : CompilerEvent | CompilerEvent[]) : string {
+  if (!e) {
+    return "";
+  }
+  if (Array.isArray(e)) {
+    return e.map(item => compilerEventFormat(item)).join(' ');
```
Comment (mcdurdin) [inline common/web/types/src/util/compiler-interfaces.ts:56]: Join with spaces and not `\n`?

File: developer/src/kmc-ldml/src/compiler/ordr.ts [hunk 6]
```diff
@@ -1,46 +1,47 @@
...
-    }
-
-    return result;
-  }
-
-  public compile(sections: GlobalSections): Ordr {
-    return this.compileReorders(sections, this.keyboard.reorders);
-  }
-}
+// import { constants } from "@keymanapp/ldml-keyboard-constants";
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/ordr.ts:1]: Should we just delete this file?

File: developer/src/kmc-ldml/src/util/pattern-parser.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
+/**
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/util/pattern-parser.ts:1]: This file is here and in common/web/types as well?

File: developer/src/kmc-ldml/test/test-bksp.ts [hunk 8]
```diff
@@ -5,23 +5,26 @@ import { compilerTestCallbacks, loadSectionFixture } from './helpers/index.js';
...
-      assert.isEmpty(bksp.items[0].before);
-      assert.strictEqual(bksp.items[0].to.value, "");
+      assert.ok(bksp);
+      // assert.lengthOf(bksp.items, 1);
+      // assert.lengthOf(bksp.items[0].from, 2);
+      // assert.strictEqual(bksp.items[0].from[0].value.value, "្");
+      // assert.strictEqual(bksp.items[0].from[1].value.value, "ម");
+      // assert.strictEqual(bksp.items[0].flags, BkspItemFlags.none);
+      // assert.isEmpty(bksp.items[0].before);
+      // assert.strictEqual(bksp.items[0].to.value, "");
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/test/test-bksp.ts:27]: TODO-LDML?

File: developer/src/kmc-ldml/test/test-dependencies.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import 'mocha';
+import {assert} from 'chai';
+// import { isValidEnumValue, calculateUniqueKeys, allUsedKeyIdsInLayers, translateLayerAttrToModifier, validModifier, verifyValidAndUnique } from '../src/util/util.js';
+// import { constants } from "@keymanapp/ldml-keyboard-constants";
+// import { LDMLKeyboard } from '@keymanapp/common-types';
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/test/test-dependencies.ts:5]:
```suggestion
```

There may be other commented code in other files that I forgot to suggest on

File: developer/src/kmc-ldml/test/test-finl.ts [hunk 10]
```diff
@@ -1,28 +1,28 @@
-import 'mocha';
-import { assert } from 'chai';
-import { FinlCompiler } from '../src/compiler/tran.js';
-import { compilerTestCallbacks, loadSectionFixture } from './helpers/index.js';
-import { KMXPlus } from '@keymanapp/common-types';
+// import 'mocha';
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/test/test-finl.ts:1]: TODO-LDML?

File: developer/src/kmc-ldml/tsconfig.json [hunk 11]
```diff
@@ -10,6 +10,7 @@
     "paths": {
       // "@keymanapp/keyman-version": ["../../../common/web/keyman-version/keyman-version.mts"],
       "@keymanapp/common-types": ["../../../common/web/types/src/main"],
+      "@keymanapp/kmc-kmn": ["../kmc-kmn/src/main"],
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/tsconfig.json:13]: I think long-term we should probably extract usetparser into its own wasm module but for now we can stick with this

**Review by mcdurdin — APPROVED** [review]
Mostly nits but one file appears to be duplicated

File: developer/src/kmc-ldml/src/util/pattern-parser.ts (hunk 7 above)
Comment (srl295) [reply developer/src/kmc-ldml/src/util/pattern-parser.ts:1]: yes, that's corrected in a downstream PR

File: common/web/types/src/kmx/kmx-plus-builder/build-vars.ts (hunk 3 above)
Comment (srl295) [reply common/web/types/src/kmx/kmx-plus-builder/build-vars.ts:28]: a typed way anyway

File: developer/src/kmc-ldml/src/util/pattern-parser.ts (hunk 7 above)
Comment (mcdurdin) [reply developer/src/kmc-ldml/src/util/pattern-parser.ts:1]: ok... for future reviews, please add a PR comment for things like this so it saves effort and mental load for reviewers 😁

File: common/web/types/src/ldml-keyboard/pattern-parser.ts (hunk 4 above)
Comment (srl295) [reply common/web/types/src/ldml-keyboard/pattern-parser.ts:58]: captured in https://github.com/keymanapp/keyman/issues/7377#issuecomment-1583569822

File: developer/src/kmc-ldml/test/test-bksp.ts (hunk 8 above)
Comment (srl295) [reply developer/src/kmc-ldml/test/test-bksp.ts:27]: parkinglot

File: developer/src/kmc-ldml/test/test-finl.ts (hunk 10 above)
Comment (srl295) [reply developer/src/kmc-ldml/test/test-finl.ts:1]: parkinglot

**Review by srl295 — COMMENTED** [review]
thanks… many suggestions copied to parking lot https://github.com/keymanapp/keyman/issues/7377#issuecomment-1583569822

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.121-alpha](https://keyman.com/downloads/releases/alpha/17.0.121)

**Pull request merged**